jbmc/src/java_bytecode/java_object_factory.cpp

```diff
@@ -756,6 +761,24 @@ void java_object_factoryt::gen_nondet_pointer_init(
     }
   }
 
+  // If this is a void* we *must* initialise with null:
+  // (This can currently happen for some cases of #exception_value)
+  bool must_be_null = subtype == empty_typet();
+
+  // If we may be about to initialize a non-null object, always run the
+  // clinit_wrapper of its class first.
+  // Note that it would be more consistent with the behaviour of the JVM to only
+  // run clinit_wrapper if we are about to initialize an object of which we know
+  // for sure that it is not null on any following branch. However, adding this
+  // case in gen_nondet_struct_init would slow symex down too much.
+  if(!must_be_null)
+  {
+    const java_class_typet &class_type = to_java_class_type(subtype);
+    const irep_idt &class_name = class_type.get_name();
+    const irep_idt class_clinit = clinit_wrapper_name(class_name);
+    gen_method_call(assignments, expr, class_clinit);
```

Comment:
Yes please @smowton ! I raised exactly this question at a few stand-ups, and @peterschrammel said it's probably best to call static initialisers for all argument types, not just enums, as that's closer to what "real Java" does. No one could really think of an example where it would really make a difference for our generated tests, or indeed why the JVM calls static initialisers before constructor calls. @peterschrammel suggested it may have something to do with concurrency, so @cesaro might have an opinion on this too.

Reply:
@antlechner But then the JVM would have to keep track of which values are statically initialised. I suspect that just keeping track of which classes have had their static initialiser run is simpler.

Reply:
Beware this might reduce coverage though if the type's static fields were unused. We may want an option to break the JVM spec in this respect and only static init upon actual static field reference.

Reply:
Good point about coverage.
@peterschrammel Should I leave this change as it is for now, and if it reduces coverage then we'll add an option for it? Or should I introduce the option right away, just in case? Or put everything behind a "if enum type"?

Reply:
Also, if we did have an option, shouldn't it apply to calling `clinit_wrapper` before regular/deterministic constructor calls too?

Reply:
> We may want an option to break the JVM spec in this respect and only static init upon actual static field reference.

We should try that to make it more a conscious choice to deviate.

Reply:
@peterschrammel As part of this PR for nondet-initialisation only, or in a future PR for both nondet and det object initialisation?